stripe/api_resources/abstract/updateable_api_resource.py

```diff
@@ -12,6 +12,11 @@ def modify(cls, sid, **params):
         return cls._static_request("post", url, params=params)
 
     def save(self, idempotency_key=None):
+        """
+        The `save` method is deprecated and will be removed in a future major version of the library.
+
+        Use the class method `modify` on the resource instead.
```

Comment:
Are there no decorators or something else in Python that will help mark the method as deprecated to tooling/editors which can then provide the appropriate warning?

Reply:
There are deprecated decorators in Python, but they would require installing a new package, and I saw that the current practice seems to just be putting it in the docstring ([`api_version` example](https://github.com/stripe/stripe-python/blob/291d569d/stripe/api_resources/file.py#L35)). I also saw that in the [last major version release](https://github.com/stripe/stripe-python/pull/844/files) the removed deprecated methods had been marked the same way).

I agree that a decorator that throws a warning would be a better user experience though. We could consider using https://deprecation.readthedocs.io/en/latest/

Reply:
You can use [`warnings.warn(message, DeprecationWarning)`](https://docs.python.org/2/library/warnings.html#warnings.warn)

Reply:
Ah yep that would be better, thank you Yejia!!

Reply:
That kind of warning shows up in run time right rather than in build time correct? Do we have precedence for that? It might be considered noisy.

If just using the docstring has been the precedent so far, then I am ok with that. We anyway have a task to set up proper deprecation workflows across languages beyond this where we can evaluate the new package

Reply:
Yes at run-time. Build-time in this case meaning linting while writing the code? I don't think we have precedence for it, but `DeprecationWarning` is ignored by default.

Agree with being ok about just using the docstring.

Reply:
The last time we used warnings was in 2019: https://github.com/stripe/stripe-python/pull/557 (description contains an example of what the warnings will look like)

But yes, this could be considered noisy. I'll revert back to docstring to align with more recent precedent.